### DISCUSSION TITLE
Using pyscf to model K-shell fluorescence

### DISCUSSION THREAD

**Discussion by willowmm** [pyscf/pyscf#2627] [Q&A]
I was hoping to use pyscf to run some calculations to compare with results from a recent experiment, in which we heated up and fluoresced samples containing Co.

Generally, I want to solve for the single-electron wave function in the Hartree Fock approximation. First, I would like to construct single-electron wave functions and use them to determine the positions of Kα and Kβ fluorescence lines in Co+, Co+2, Co+3, etc. This could be done with Gaussian-type orbitals. 

Next, I would like to introduce an assumption about the surrounding environment in the wavefunction calculation. The energy levels of an ion immersed in a plasma environment are obtained from the Schrodinger equation for the ion in a medium (plasma environment):
 
$E_n \psi_n = (H_{ion} + V_{plasma}(r)) \psi_n$
 
We consider the interaction with the plasma V_plasma(r) as a perturbation and determine first the eigenstates of the isolated ion for different configurations. Energy levels and wavefunctions are approximated by the solution of the self-consistent Roothaan-Hartree-Fock (RHF) equations for unperturbed atom or ion, respectively.
 
The shifts of energy levels of an ion in a plasma environment is estimated within a self-consistent ion sphere model, where the radius of the ion sphere is given by
 
$R = (\frac{3 Z_{ion}}{4 \pi n_e^0 })^{1/3}$
 
Where Z_ion is the ionic charge (the sum of the nuclear charge and the charge of the bound electrons). I want to consider the effect of the polarization of free electrons by calculating wave functions under the assumption that the Co ion is immersed in an effective mean field of free electrons with spatially-averaged density $n_e^0$.
 
I want to calculate the Hartree-Fock potential including all bound states. The density distribution n_e(r) can be separated into components from bound and free electrons. Assuming only one ion per sphere, the total potential $\phi(r)$ within the sphere can be determined via the Poisson equation:

$\Delta \phi(r) = 4 \pi n_{free}(r) + 4 \pi e n_{bound}(r) - 4 \pi Z e \delta(r)$
$\phi(r) = \phi_{free}(r) + \phi_{bound}(r) + \frac{Z e}{r}$
$V_{plasma}(r) = - e \phi_{free}(r)$
 
and the free and bound electron densities are given by:
 
$n_{free}(r) = n_{free}(R) \frac{4}{\sqrt{\pi}} \int_{p_0 = \sqrt{2 m_e e \phi(r)}}^{\infty} dp \frac{p^2}{(2 m k_b T)^{3/2}} e^{-\frac{1}{k_b T} (\frac{p^2}{2 m} - e \phi(r))}$
$n_{bound}(r) = \sum_{nl}^{(bound)} \frac{\| R_{nl}(r)\|^2}{4 \pi}$
 
To find an iterative solution for the plasma potential we could start with a uniform free electron distribution of density $n_e^0$.
 
After calculating the perturbing plasma potential, I would like to use it to caulate the energy level shifts it generates:
 
$\Delta E_{i,f} = \braket{\psi_i | V_{plasma}(r) | \psi_i} - \braket{\psi_f | V_{plasma}(r) | \psi_f}$
 
Before I even move on to evaluate the plasma potential, one nagging question I have though is about how evaluate the expected K-shell fluorescence energies even for an atom / ion in isolation. I wrote the following code attempting to manually enforce that there is only a single electron in the K-shell, but it seems like the calculation must be very inaccurate. The K-alpha and K-beta fluorescence energies should be ~ 6.9 keV and 7.6 keV respectively, (~0.7 eV apart from one another), but the energies I calculate right now for each are only separated by ~ 0.1 eV). This holds whether I apply the procedure to customize the occupation of the K-shell. I will post the code I wrote to make these calculations below:

```
from pyscf import gto, scf
import numpy as np
 
eV_per_Hartree = 27.2114
 
# Function to compute orbital energies for a given ion with custom K-shell occupation
def compute_ion_spectrum(atom, atomic_number, charge):
    num_electrons = atomic_number - charge
    spin = (num_electrons % 2)
    mol = gto.M(atom=f'{atom} 0 0 0', charge=charge, spin=spin, basis='Def2-SVP')
    mf = scf.RHF(mol)
 
    # Custom occupation: Partial ionization of the K-shell
    def custom_occ(mo_energy, mo_occ, *args):
        num_orbitals = len(mo_energy)  # Number of molecular orbitals
        mo_occ = np.zeros(num_orbitals)  # Initialize all occupations to 0
        if num_orbitals < 5:
            raise ValueError("Basis set or configuration has too few orbitals!")
        mo_occ[0] = 1  # Single electron in the 1s orbital (K-shell)
        zbar = atomic_number - charge
        mo_occ[1:(zbar//2)] = 2  # Fully occupy the next shells according to the Aufbau principle
        total_occupancy = np.sum(mo_occ)  # Total number of occupied electrons
        if total_occupancy > num_electrons:
            raise ValueError(
                f"Total occupancy ({total_occupancy}) exceeds available electrons ({num_electrons})!"
            )
        return mo_occ
 
    # Apply the custom occupation to the SCF object
    mf.get_occ = custom_occ
 
    # Run the SCF calculation
    try:
        mf.kernel()
    except Exception as e:
        print(f"SCF calculation failed for charge={charge}, error: {e}")
        return []
 
    # Get orbital energies
    orbital_energies = mf.mo_energy
    return orbital_energies
 
# Define the ions to compute
charges = [1, 6, 12]  # Co+, Co6+, Co12+
 
# Compute Kα and Kβ lines
for charge in charges:
    print(f"\nComputing for Co^{charge}+...")
    energies = compute_ion_spectrum('Co', 27, charge)
    if len(energies) < 2:
        print("Insufficient data to calculate Kα and Kβ lines.")
        continue
    # Assuming energy levels are sorted, extract relevant transitions:
    # 1s -> 2p (Kα) and 1s -> 3p (Kβ)
    k_alpha = energies[1] - energies[0]  # Energy difference between 2p and 1s
    k_beta = energies[2] - energies[0]   # Energy difference between 3p and 1s
    k_alpha *= eV_per_Hartree
    k_beta *= eV_per_Hartree
    print(f"Co^{charge}+: Kα = {k_alpha:.4f} eV, Kβ = {k_beta:.4f} eV")
 
# This code is producing results that are totally off; Co Kb only separated by 110 eV, when it should be off by more like 719 eV
```

In summary, I have two immediate questions: Have I misunderstood how the orbitals are being indexed? And is it possible to customize the occupation of specific orbitals before evaluating their energies?

If someone is able to help me with the above issues, I would also request guidance in how to best calculate a self-consistent field model in the manner I describe above. Thank you so much!

**Comment by sunqm** [comment]
You can customize the orbital occupancy this way, forcing the lowest-energy orbital to be singly occupied. However, I’m not sure if this is meaningful for the RHF method, as it requires orbitals to be doubly occupied. UHF might be a more appropriate option when dealing with non-symmetric occupation between spin-up and spin-down electrons.
